Framework/Algorithms/src/CylinderAbsorption.cpp

```diff
@@ -99,46 +149,27 @@ std::string CylinderAbsorption::sampleXML() {
 /// Calculate the distances for L1 and element size for each element in the
 /// sample
 void CylinderAbsorption::initialiseCachedDistances() {
-  m_L1s.resize(m_numVolumeElements);
-  m_elementVolumes.resize(m_numVolumeElements);
-  m_elementPositions.resize(m_numVolumeElements);
-
-  int counter = 0;
-  // loop over slices
-  for (int i = 0; i < m_numSlices; ++i) {
-    const double z = (i + 0.5) * m_sliceThickness - 0.5 * m_cylHeight;
-
-    // Number of elements in 1st annulus
-    int Ni = 0;
-    // loop over annuli
-    for (int j = 0; j < m_numAnnuli; ++j) {
-      Ni += 6;
-      const double R = (j * m_cylRadius / m_numAnnuli) + (m_deltaR / 2.0);
-      // loop over elements in current annulus
-      for (int k = 0; k < Ni; ++k) {
-        const double phi = 2 * M_PI * k / Ni;
-        // Calculate the current position in the sample in Cartesian
-        // coordinates.
-        // Remember that our cylinder has its axis along the y axis
-        m_elementPositions[counter](R * sin(phi), z, R * cos(phi));
-        assert(m_sampleObject->isValid(m_elementPositions[counter]));
-        // Create track for distance in cylinder before scattering point
-        Track incoming(m_elementPositions[counter], m_beamDirection * -1.0);
-
-        m_sampleObject->interceptSurface(incoming);
-        m_L1s[counter] = incoming.cbegin()->distFromStart;
-
-        // Also calculate element volumes here
-        const double outerR = R + (m_deltaR / 2.0);
-        const double innerR = outerR - m_deltaR;
-        const double elementVolume =
-            M_PI * (outerR * outerR - innerR * innerR) * m_sliceThickness / Ni;
-        m_elementVolumes[counter] = elementVolume;
-
-        counter++;
-      }
-    }
-  }
+  if (!m_sampleObject) // should never happen
+    throw std::runtime_error("Do not have a sample object defined");
+
+  if (m_sampleObject->shape() !=
+      Geometry::detail::ShapeInfo::GeometryShape::CYLINDER)
+    throw std::runtime_error("Sample shape is not a cylinder");
+  const Geometry::CSGObject *shape =
+      dynamic_cast<const Geometry::CSGObject *>(m_sampleObject);
+  if (!shape)
+    throw std::runtime_error(
+        "Failed to convert shape from IObject to CSGObject");
+  const auto raster = Geometry::Rasterize::calculateCylinder(
+      m_beamDirection, *shape, m_numSlices, m_numAnnuli);
+  m_sampleVolume = raster.totalvolume;
+  if (raster.l1.size() == 0)
+    throw std::runtime_error("Failed to rasterize shape");
+  // move over the information
+  m_numVolumeElements = raster.l1.size();
+  m_L1s = std::move(raster.l1);
```

Comment:
`raster` is still marked as `const`. I don't think `std::move` will do what you think here.